File: src/System/Notification/Cocoa.hs
{-# LANGUAGE DataKinds, OverloadedStrings, QuasiQuotes, RecordWildCards #-}
{-# LANGUAGE TemplateHaskell                                            #-}
module System.Notification.Cocoa
    ( Notification (..)
    , newNotification
    , defaultNotificationSound
    , deliver
    , schedule
    ) where

import           Control.Monad         (when)
import           Data.ByteString       (ByteString)
import           Data.Maybe            (fromJust, fromMaybe, isJust)
import           Data.Monoid           ((<>))
import           Data.String           (IsString (fromString))
import           Data.Text             (Text)
import qualified Data.Text             as T
import           Data.Time             (DiffTime, ZonedTime, timeZoneMinutes)
import           Data.Time             (zonedTimeToUTC, zonedTimeZone)
import           Data.Time             (getCurrentTime)
import           Data.Time             (utcToZonedTime)
import           Data.Time             (getCurrentTimeZone)
import           Data.Time.Clock.POSIX (utcTimeToPOSIXSeconds)
import           Foreign.C             (CDouble (..), CInt (..), CLong (..))
import           Language.ObjC.Inline  (ObjC, block, defClass)
import           Language.ObjC.Inline  (objcCtxWithClasses)
import           Language.ObjC.Inline  (toObjC')
import qualified Language.ObjC.Inline  as C
import           System.Random         (randomIO)

type NSUserNotification = ObjC "NSUserNotification"
type NSUserNotificationCenter = ObjC "NSUserNotificationCenter"

C.context (objcCtxWithClasses [defClass "NSUserNotificationCenter"
                              ,defClass "NSUserNotification"
                              ])
C.import_ "<Foundation/Foundation.h>"
C.import_ "BundleSnatcher.h"

data Notification = Notification
                    { notifTitle                  :: Text
                    , notifSubtitle               :: Maybe Text
                    , notifInformativeText        :: Text
                    , notifImage                  :: Maybe ByteString
                    , notifIdentifier             :: Maybe Text
                    , notifActionButton           :: Maybe Text
                    , notifOtherButton            :: Maybe Text
                    , notifHasReplyButton         :: Bool
                    , notifDeliveryDate           :: Maybe ZonedTime
                    , notifDeliveryRepeatInterval :: Maybe DiffTime
                    , notifSoundName              :: Maybe Text
                    , notifBundleName             :: Text
                    }
                  deriving (Show)

newNotification :: Notification
newNotification = Notification { notifTitle = ""
                               , notifSubtitle = Nothing
                               , notifInformativeText = ""
                               , notifImage = Nothing
                               , notifIdentifier = Nothing
                               , notifActionButton = Nothing
                               , notifOtherButton = Nothing
                               , notifHasReplyButton = False
                               , notifDeliveryDate = Nothing
                               , notifDeliveryRepeatInterval = Nothing
                               , notifSoundName  = Nothing
                               , notifBundleName = "com.konn-san.dummy-notifier"
                               }

instance IsString Notification where
  fromString str = newNotification { notifTitle = T.pack str }

defaultCenter :: NSUserNotificationCenter
defaultCenter = [C.pure|NSUserNotificationCenter * {
                         [NSUserNotificationCenter defaultUserNotificationCenter]
                        }|]

newUserNotification :: Notification -> IO NSUserNotification
newUserNotification Notification{..} = do
  i <- randomIO
  let subt = fromMaybe "" notifSubtitle
      hasAct = fromIntegral $ fromEnum $ isJust notifActionButton
      hasRepl = fromIntegral $ fromEnum notifHasReplyButton
      actBName = fromMaybe "" notifActionButton
      othBName = fromMaybe "" notifOtherButton
      ident = fromMaybe ("haskell-cocoa-notif-" <> T.pack (show (i :: Int))) notifIdentifier
  sound <- maybe [C.exp|NSString * { nil }|] toObjC' notifSoundName
  notif <- [block| NSUserNotification * {
                 NSUserNotification *notif = [NSUserNotification new];
                 notif.title = $obj:(NSString *notifTitle);
                 notif.subtitle  = $obj:(NSString *subt);
                 notif.informativeText = $obj:(NSString *notifInformativeText);
                 notif.identifier = $obj:(NSString *ident);
                 notif.hasActionButton = $(BOOL hasAct);
                 notif.actionButtonTitle = $obj:(NSString *actBName);
                 notif.otherButtonTitle = $obj:(NSString *othBName);
                 notif.hasReplyButton = $(BOOL hasRepl);
                 notif.soundName = $raw:(NSString *sound);
                 return notif;
               } |]
  when (isJust notifDeliveryDate) $ do
    let date = fromJust notifDeliveryDate
        zone = toEnum $ timeZoneMinutes (zonedTimeZone date) * 60
        secs = fromRational $ toRational $ utcTimeToPOSIXSeconds $ zonedTimeToUTC date
    [block| void {
              NSUserNotification *notif = $raw:(NSUserNotification *notif);
              notif.deliveryDate = [NSDate dateWithTimeIntervalSince1970: $(double secs)];
              NSLog(@"deliv Date set");
              notif.deliveryTimeZone = [NSTimeZone timeZoneForSecondsFromGMT: $(NSInteger zone)];
              NSLog(@"deliv Zone set");
            } |]
  return notif

defaultNotificationSound :: Text
defaultNotificationSound = [pure'| NSString * { NSUserNotificationDefaultSoundName } |]


deliver :: Notification -> IO ()
deliver notif = do
  putStrLn "initing..."
  obj <- newUserNotification notif
  putStrLn "inited. passing..."
  [C.exp| void { [[NSUserNotificationCenter defaultUserNotificationCenter] deliverNotification: $raw:(NSUserNotification *obj)] } |]

schedule :: Notification -> IO ()
schedule notif = do
  putStrLn "initing..."
  zone <- getCurrentTimeZone
  now <- utcToZonedTime zone <$> getCurrentTime
  let notif' | isJust (notifDeliveryDate notif) = notif
             | otherwise = notif { notifDeliveryDate = Just now }
  obj <- newUserNotification notif'
  let name = notifBundleName notif
  putStrLn "inited. passing..."
  [C.exp| void {
              snatchingBundle($obj:(NSString *name),
                              ^void {[[NSUserNotificationCenter defaultUserNotificationCenter] scheduleNotification: $raw:(NSUserNotification *obj)];}) } |]

File: src/System/Notification/Cocoa.m

#import <Foundation/Foundation.h>

#import "BundleSnatcher.h"

NSUserNotificationCenter * inline_c_System_Notification_Cocoa_0_1372577b584c30ce3627c4a7dedfec27ca068c60() {
return (
                         [NSUserNotificationCenter defaultUserNotificationCenter]
                        );
}


NSString * inline_c_System_Notification_Cocoa_1_4d49a7bf918b0d45f9ef6565fd96773943ccad17() {
return ( nil );
}


NSUserNotification * inline_c_System_Notification_Cocoa_2_7ef372a21e3564fc0098b711964efe9fed7a25e5(NSString * notifTitle_inline_c_0, NSString * subt_inline_c_1, NSString * notifInformativeText_inline_c_2, NSString * ident_inline_c_3, BOOL hasAct_inline_c_4, NSString * actBName_inline_c_5, NSString * othBName_inline_c_6, BOOL hasRepl_inline_c_7, NSString * sound_inline_c_8) {

                 NSUserNotification *notif = [NSUserNotification new];
                 notif.title = notifTitle_inline_c_0;
                 notif.subtitle  = subt_inline_c_1;
                 notif.informativeText = notifInformativeText_inline_c_2;
                 notif.identifier = ident_inline_c_3;
                 notif.hasActionButton = hasAct_inline_c_4;
                 notif.actionButtonTitle = actBName_inline_c_5;
                 notif.otherButtonTitle = othBName_inline_c_6;
                 notif.hasReplyButton = hasRepl_inline_c_7;
                 notif.soundName = sound_inline_c_8;
                 return notif;
               
}


void inline_c_System_Notification_Cocoa_3_934447d7ea3cff821d33e46098edab39fdb93ac2(NSUserNotification * notif_inline_c_0, double secs_inline_c_1, NSInteger zone_inline_c_2) {

              NSUserNotification *notif = notif_inline_c_0;
              notif.deliveryDate = [NSDate dateWithTimeIntervalSince1970: secs_inline_c_1];
              NSLog(@"deliv Date set");
              notif.deliveryTimeZone = [NSTimeZone timeZoneForSecondsFromGMT: zone_inline_c_2];
              NSLog(@"deliv Zone set");
            
}


void inline_c_System_Notification_Cocoa_4_51df0313d192e642c664b96bdf023522eeab36e1(NSUserNotification * obj_inline_c_0) {
 [[NSUserNotificationCenter defaultUserNotificationCenter] deliverNotification: obj_inline_c_0] ;
}


void inline_c_System_Notification_Cocoa_5_707c0af0faa1856ed12a2c4a50f55f20d95c0110(NSString * name_inline_c_0, NSUserNotification * obj_inline_c_1) {

              snatchingBundle(name_inline_c_0,
                              ^void {[[NSUserNotificationCenter defaultUserNotificationCenter] scheduleNotification: obj_inline_c_1];}) ;
}

